File: core/src/main/scala/tripaint/coords/TriangleCoords.scala
package tripaint.coords

case class TriangleCoords(value: Int) extends AnyVal {
  inline def x: Int = value >> 12
  inline def y: Int = value & 0xfff

  def toInt: Int = x << 12 | y
}

object TriangleCoords {
  inline def apply(x: Int, y: Int): TriangleCoords = {
    assert(x >= 0)
    assert(x <= 2 * y)
    assert(y < 0x1000)
    new TriangleCoords(x << 12 | y)
  }

  def fromInt(repr: Int): TriangleCoords = {
    assert(repr != -1)
    TriangleCoords(repr >>> 12, repr & 0xfff)
  }
}

File: core/src/main/kotlin/tripaint/coords/TriangleCoords.kt
package tripaint.coords

data class TriangleCoords(val value: Int) {
    val x: Int
        get() = value shr 12
    val y: Int
        get() = value and 0xfff

    fun toInt(): Int = (x shl 12) or y

    companion object {
        fun from(x: Int, y: Int): TriangleCoords {
            assert(x >= 0)
            assert(x <= 2 * y)
            assert(y < 0x1000)
            return TriangleCoords((x shl 12) or y)
        }

        fun fromInt(repr: Int): TriangleCoords {
            assert(repr != -1)
            return from(repr ushr 12, repr and 0xfff)
        }
    }

}